### PR TITLE
Add some clarity to contributors file

### PR DESCRIPTION
**Pull request by bahiirwa**
Following the core meeting https://forums.classicpress.net/t/april-2021-core-meeting/3114/3

Some issues where raised about the contributors.md file. This is an initial step to reduce the contributor's challenges in reading and following the document.

Changes
- [x] Adds elaborate use of the bash backport script.
- [x] Shorter sentences for clarity.
- [x] Updated links to the forums and slack channel.
- [ ] Add video showing a simple backport.

### CODE REVIEW

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -1,6 +1,8 @@
 # Contributions
 
-ClassicPress is a volunteer-driven open source project that thrives when we have multiple contributors working towards a common goal.
+ClassicPress is a volunteer-driven open source project that thrives when we have multiple contributors working towards a common goal. 
+
+There may be some delays in reviewing and merging pull requests. This is because we take a lot of care with our platform and ClassicPress is maintained by volunteers in our free time.
```
Comment (nylen) [inline .github/CONTRIBUTING.md:5]: How about just removing this sentence instead? The delay for reviewing PRs should be shortened when we have more people looking at them

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -13,46 +15,45 @@ Also, please be sure to follow our [code of conduct](https://www.classicpress.ne
...
+  - Review and test [existing PRs](https://github.com/ClassicPress/ClassicPress/pulls), especially looking at how well they fit the criteria described above. Let us know how you tested the changes and what you found. Screenshots, gifs, video and text instructions documenting the tests can be very useful. We need to be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better.
   - Bugfixes, you can find a list of known bugs on our [issues page](https://github.com/ClassicPress/ClassicPress/issues).
   - Issues with the [`help wanted`](https://github.com/ClassicPress/ClassicPress/labels/help%20wanted) or [`good first issue`](https://github.com/ClassicPress/ClassicPress/labels/good%20first%20issue) labels.
-  - [Planned](https://petitions.classicpress.net/?view=planned) or [most wanted](https://petitions.classicpress.net/?view=most-wanted) feature requests from our petitions site.
```
Comment (nylen) [inline .github/CONTRIBUTING.md:40]: Why was this part removed?

File: .github/CONTRIBUTING.md [hunk 3]
```diff
@@ -71,22 +72,44 @@ Documentation for running and updating our existing tests, as well as the code f
...
 
-Otherwise, if you're interested in backporting a change from WordPress, then please explain in a pull request or issue what the specific change is, how it works, and why it's something that makes sense for ClassicPress to adopt. Often there is a lot of good discussion about each change in the relevant WordPress Trac tickets, and it makes the maintainers' job much easier to see this discussion summarized and linked.
+### Tips for a good Backport PR.
+- Give the PRs a clear title. Explain in a pull request or issue what the specific change is, how it works and why ClassicPress in particular should adopt it. Often there is a lot of good discussion about each change in the relevant WordPress Trac tickets. It makes the maintainers' job much easier to see this discussion summarized and linked.
+- Explain what testing has been done on the PR and what may be left to do. Screenshots, gif, video and text instructions documenting the tests can be very useful. Those tell other contributors looking at the PR how to verify the changes.
```
Comment (nylen) [inline .github/CONTRIBUTING.md:87]:
```suggestion
- Explain what testing has been done on the PR and what may be left to do. Screenshots, gif, video and text instructions documenting the tests can be very useful. Those tell other contributors looking at the PR how to verify the changes. "This was tested in WordPress" **is not enough** to meet our review criteria, see [Review Criteria](#review-criteria) above.
```

File: .github/CONTRIBUTING.md [hunk 4]
```diff
@@ -71,22 +72,44 @@ Documentation for running and updating our existing tests, as well as the code f
...
+Then push your branch for review as a PR.
 
-Using this script for all backports saves time for you and for the maintainers, and it also uses a standardized format for commit messages which makes it possible for us to track which WordPress changes we've already included.
+Using this script for all backports saves time for you and for the maintainers. As it uses a standardized format for commit messages making it possible for us to track which WordPress changes we've already included.
 
-If there are merge conflicts for your changes, the script will let you know. Please fix the conflicts **in a separate commit** so that we can take a look at the conflicts and their fixes, or if you're not comfortable resolving merge conflicts, just push your branch as-is and we'll take care of it!
+**Note:** If there are merge conflicts for your changes, the script will let you know. Please fix the conflicts **in a separate commit** so that we can take a look at the conflicts and their fixes, or if you're not comßfortable resolving merge conflicts, just push your branch as-is and we'll take care of it!
```
Comment (nylen) [inline .github/CONTRIBUTING.md:113]: Extra non-English letter inserted here:

```suggestion
**Note:** If there are merge conflicts for your changes, the script will let you know. Please fix the conflicts **in a separate commit** so that we can take a look at the conflicts and their fixes, or if you're not comfortable resolving merge conflicts, just push your branch as-is and we'll take care of it!
```

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -71,22 +72,44 @@ Documentation for running and updating our existing tests, as well as the code f
...
+```
+bin/backport-wp-commit.sh -c "changeset-number"
+```
+to apply all the changesets to the same branch. 
 
-Finally, when you're ready to backport a code change, identify the WordPress **changeset number** that you'd like to port, and run the `bin/backport-wp-commit.sh` script to apply the change to the code. If you're porting multiple changesets, you can use the `-c` option to this script to apply all the changesets to the same branch.
+Then push your branch for review as a PR.
 
-Using this script for all backports saves time for you and for the maintainers, and it also uses a standardized format for commit messages which makes it possible for us to track which WordPress changes we've already included.
+Using this script for all backports saves time for you and for the maintainers. As it uses a standardized format for commit messages making it possible for us to track which WordPress changes we've already included.
```
Comment (nylen) [inline .github/CONTRIBUTING.md:111]: Fine to put this in two separate sentences but grammar is a bit off.

```suggestion
Using this script for all backports saves time for you and for the maintainers. It uses a standardized format for commit messages, which makes it possible for us to track which WordPress changes we've already included.
```

File: .github/CONTRIBUTING.md [hunk 6]
```diff
@@ -13,46 +15,45 @@ Also, please be sure to follow our [code of conduct](https://www.classicpress.ne
...
   - Bugfixes, you can find a list of known bugs on our [issues page](https://github.com/ClassicPress/ClassicPress/issues).
   - Issues with the [`help wanted`](https://github.com/ClassicPress/ClassicPress/labels/help%20wanted) or [`good first issue`](https://github.com/ClassicPress/ClassicPress/labels/good%20first%20issue) labels.
-  - [Planned](https://petitions.classicpress.net/?view=planned) or [most wanted](https://petitions.classicpress.net/?view=most-wanted) feature requests from our petitions site.
-  - Exploratory PRs with your own suggested changes, but please remember these will be subject to review to make sure they are in line with the project's direction.
+  - [Planned Milestones](https://github.com/ClassicPress/ClassicPress/milestones) or exploratory PRs with your own suggested changes. Please remember these will be subject to review to make sure they are in line with the project's direction.
```
Comment (nylen) [inline .github/CONTRIBUTING.md:40]: I think adding milestones is good but it should probably go with the sentence above instead (the one that was removed).

**Comment by nylen** [comment]
Thanks for the PR!

I left some questions and suggestions.

One thing that's missing is some quick instructions for how to create a ClassicPress development environment.

File: .github/CONTRIBUTING.md (hunk 1 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md:5]: 50/50 for me. One of the most unexpected things in FOSS is people assuming everything is snap. This is a gentle reminder. But we can agree to disagree.

File: .github/CONTRIBUTING.md (hunk 2 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md:40]: We have since moved the petitions to forums. We haven't yet found a way to show the most wanted and planned. Milestones on Github might be the next best thing to share this until we figure the forums route out.

File: .github/CONTRIBUTING.md (hunk 6 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md:40]: I guess we can do that for now until we retire the petitions subdomain. However, it now says, "it is archived." All the petitions are on the forums. We are using "hearts" to poll until a properway to do this is chosen. Would you still suggest we keep pointing to this?

File: .github/CONTRIBUTING.md (hunk 2 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md:40]: Let's keep the conversation to this via https://github.com/ClassicPress/ClassicPress/pull/710#discussion_r619798985

**Comment by bahiirwa** [comment]
Thanks for the review @nylen. I have left some rebuttals and commits on suggestions.

> One thing that's missing is some quick instructions for how to create a ClassicPress development environment, and we could probably say a bit more about reviewing PRs.

Agreed. 

1. Can we make this a new issue to keep this a small PR for review? I also think this should go to the main `readme.md`. It is essential to see when one wants to test CP on the fly even without need to give feedback or contribute.
2. I would request for your expertise for the suggestion.

**Comment by nylen** [comment]
>     1. Can we make this a new issue to keep this a small PR for review?

Yep sounds good to me. We can discuss the local dev instructions at #712.

File: .github/CONTRIBUTING.md (hunk 6 above)
Comment (nylen) [reply .github/CONTRIBUTING.md:40]: OK, I see why that was removed. The archive domain will stay up (to keep old links working, point people to the new location, and show a better view of what people decided they wanted to see in ClassicPress), but it will become more outdated over time. Also, jumping right in with those petitions is probably not the best place to get started with contributing.

File: .github/CONTRIBUTING.md (hunk 1 above)
Comment (nylen) [reply .github/CONTRIBUTING.md:5]: Maybe we can make the reminder more gentle then:

```suggestion
We invite all contributors to not only submit pull requests that are in line with the goals and direction of the project, but also to discuss and review other contributors' issues and pull requests. This will help us make the most of our limited time and review all contributions quickly.
```

File: .github/CONTRIBUTING.md (hunk 6 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md:40]: I think that resolves it.

File: .github/CONTRIBUTING.md [hunk 7]
```diff
@@ -71,22 +72,44 @@ Documentation for running and updating our existing tests, as well as the code f
...
-If you're not sure about any of that, then it's a good idea to ask first. A good way is to create an issue for the specific change you're interested in, along with links to the relevant WordPress changesets and tickets, and any other information you have about how the change works.
+### Making a Backport PR.
+When you're ready to backport a code change, make a new branch to your ClassicPress fork:
+
+- Identify the WordPress **changeset number** that you'd like to port such as `43123`.
+- Run `bin/backport-wp-commit.sh` script in your terminal/command prompt.
+
+```
+bin/backport-wp-commit.sh "changeset-number"
+```
```
Comment (joyously) [inline .github/CONTRIBUTING.md:99]: Are the quotes needed? Can you put the actual number in the command?

File: .github/CONTRIBUTING.md [hunk 8]
```diff
@@ -71,22 +72,44 @@ Documentation for running and updating our existing tests, as well as the code f
...
+ to apply the change to the code. 
+ 
+ If you're porting multiple changesets, you can use the `-c` option to this script 
+
+```
+bin/backport-wp-commit.sh -c "changeset-number"
```
Comment (joyously) [inline .github/CONTRIBUTING.md:105]: This example is for multiple changesets, so please show the correct syntax.

File: .github/CONTRIBUTING.md (hunk 7 above)
Comment (nylen) [reply .github/CONTRIBUTING.md:99]: The changeset number is the number of the WP changeset to backport. The user of the script will put the actual number in the command.

I usually use the format `CHANGESET_NUMBER` for this. Quotes are probably best avoided since they have special meaning on the command line.

File: .github/CONTRIBUTING.md (hunk 8 above)
Comment (nylen) [reply .github/CONTRIBUTING.md:105]: The syntax is correct (changesets must be backported one at a time), but it can be explained better that to backport multiple changesets you do the following:

1. Start a new branch
2. `bin/backport-wp-commit.sh -c CHANGESET_NUMBER`
3. Resolve merge conflicts (if any)
4. Repeat steps 2 and 3

File: .github/CONTRIBUTING.md [hunk 9]
```diff
@@ -71,22 +72,44 @@ Documentation for running and updating our existing tests, as well as the code f
...
+- Look to make sure the backport is done correctly, with attribution to the original WP committer. 
+- If there are merge conflicts during the backport, check if they have been resolved correctly by comparing the final changes from the full PR against the corresponding WP changeset(s). Remember to fix these **in a separate commit**. If you cannot resolve the conflicts, ask for help in the [Core slack channel](https://www.classicpress.net/join-slack/) or just push your branch as-is and we'll take care of it!
 
-If you're not sure about any of that, then it's a good idea to ask first. A good way is to create an issue for the specific change you're interested in, along with links to the relevant WordPress changesets and tickets, and any other information you have about how the change works.
+### Making a Backport PR.
+When you're ready to backport a code change, make a new branch to your ClassicPress fork:
```
Comment (joyously) [inline .github/CONTRIBUTING.md:92]: > make a new branch to your ClassicPress fork:

This sounds like the next thing will show how to make a branch in the fork, but the next thing is about identifying a changeset (sort of).  Does the script make a branch? If so, say so. If not, put the command to make the branch.

File: .github/CONTRIBUTING.md [hunk 10]
```diff
@@ -71,22 +72,44 @@ Documentation for running and updating our existing tests, as well as the code f
...
+ 
+ If you're porting multiple changesets, you can use the `-c` option to this script 
+
+```
+bin/backport-wp-commit.sh -c "changeset-number"
+```
+to apply all the changesets to the same branch. 
 
-Finally, when you're ready to backport a code change, identify the WordPress **changeset number** that you'd like to port, and run the `bin/backport-wp-commit.sh` script to apply the change to the code. If you're porting multiple changesets, you can use the `-c` option to this script to apply all the changesets to the same branch.
+Then push your branch for review as a PR.
```
Comment (joyously) [inline .github/CONTRIBUTING.md:109]: Say how to push your branch. At least mention it's your fork, not the ClassicPress repo.

**Comment by bahiirwa** [comment]
Thanks @joyously.

I had not see some the conversation above. However, I have made some improvements.

- [x] Added a Table of contents as the document is becoming massive. This will help with the navigation.
- [x] Added sections for reviewing PR and Setting up a local environment.

I would appreciate your good eyes on this. 

@nylen if there are more productive ways to git some of those things. I would appreciate some correction.

File: .github/CONTRIBUTING.md (hunk 10 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md:145]: Agree this might have to change with the new section of making your local dev environment. Let's fix them hand in hand.

File: .github/CONTRIBUTING.md (hunk 9 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md:128]: https://github.com/ClassicPress/ClassicPress/pull/710#discussion_r619825795

**Comment by nylen** [comment]
I will make another pass at this later. Are we doing all of #712 in this PR then?

File: .github/CONTRIBUTING.md (hunk 8 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md]: Made some changes

**Comment by bahiirwa** [comment]
Yes. I think with the ToC, we can make one mega change.

File: .github/CONTRIBUTING.md (hunk 7 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md:140]: Made changes and used a block letter variable.

**Comment by bahiirwa** [comment]
Quick one: Is the .github/images/ the best place to store the images going to the contributors document?

**Comment by bahiirwa** [comment]
Should we include this as a link https://github.com/ClassicPress/ClassicPress/issues/692#issuecomment-821882221 as well for those who might be looking to fixing the CS? 

It is a nicely done.

**Comment by nylen** [comment]
I will make another pass at this later. Are we doing all of #712 in this PR then?

> Quick one: Is the `.github/images/` the best place to store the images going to the contributors document?

Yes, this is fine.

> Should we include [link] as well for those who might be looking to fixing the CS?

I don't think so, fixing coding styles should be handled separately and people will already need some experience with contributing to ClassicPress before doing this.

Also, having clean code will be nice, but honestly, we have much more urgent things to focus on.

File: .github/CONTRIBUTING.md [hunk 11]
```diff
@@ -1,58 +1,95 @@
...
 
 Before contributing to this repository, please read our [Democracy page](https://www.classicpress.net/democracy/) to understand how we make decisions about what is included in ClassicPress, and our [roadmap](https://www.classicpress.net/roadmap/) to see what's already planned for the next couple of releases.
 
 The **petitions process** mentioned in our Democracy document is very important to help us prioritize new features based on the needs of our users. There are some [exceptions](https://www.classicpress.net/democracy/#democracy-exceptions) to this process for minor changes and bugfixes, but generally speaking it is a good idea to search or ask in one of our communication channels (see below) before undertaking a change, because most changes should go through the petitions process.
 
 Also, please be sure to follow our [code of conduct](https://www.classicpress.net/democracy/#democracy-conduct) in all interactions with ClassicPress community members.
 
+## Quick Links
```
Comment (nylen) [inline .github/CONTRIBUTING.md:13]:
```suggestion
## Table of Contents

```

"Quick Links" makes it sound like this goes out to another part of the document.

File: .github/CONTRIBUTING.md [hunk 12]
```diff
@@ -1,58 +1,95 @@
...
+  - [Review criteria](#review-criteria)
+  - [What to work on?](#what-to-work-on)
+  - [Setting up Local Testing and Dev Environment](#setting-up-local-testing-and-dev-environment)
+  - [More information](#more-information)
+    - [Tips for good Pull Requests (PRs)](#tips-for-good-pull-requests-prs)
+    - [Automated tests](#automated-tests)
+    - [Backporting changes from WordPress](#backporting-changes-from-wordpress)
+    - [Tips for a good Backport PR.](#tips-for-a-good-backport-pr)
+    - [Making a Backport PR.](#making-a-backport-pr)
+    - [How to review a PR](#how-to-review-a-pr)
```
Comment (nylen) [inline .github/CONTRIBUTING.md:26]: No need to link to the main heading or the table of contents itself.

```suggestion
- [Communication channels](#communication-channels)
- [Review criteria](#review-criteria)
- [What to work on?](#what-to-work-on)
- [Setting up Local Testing and Dev Environment](#setting-up-local-testing-and-dev-environment)
- [More information](#more-information)
  - [Tips for good Pull Requests (PRs)](#tips-for-good-pull-requests-prs)
  - [Automated tests](#automated-tests)
  - [Backporting changes from WordPress](#backporting-changes-from-wordpress)
  - [Tips for a good Backport PR.](#tips-for-a-good-backport-pr)
  - [Making a Backport PR.](#making-a-backport-pr)
  - [How to review a PR](#how-to-review-a-pr)
```

File: .github/CONTRIBUTING.md [hunk 13]
```diff
@@ -1,58 +1,95 @@
...
 If your change meets all of these criteria then we will most likely accept it.
 
-We take a lot of care with our platform, and ClassicPress is maintained by volunteers in our free time. This means there may be some delays in reviewing and merging pull requests.
-
 ## What to work on?
 
 If you're not sure where to start contributing, here are some ideas:
 
-  - Review and test [existing PRs](https://github.com/ClassicPress/ClassicPress/pulls), especially looking at how well they fit the criteria described above. Let us know how you tested the changes and what you found. We need to be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better.
+  - Review and test [existing PRs](https://github.com/ClassicPress/ClassicPress/pulls), especially looking at how well they fit the criteria described above. Let us know how you tested the changes and what you found. Screenshots, gifs, video and text instructions documenting the tests can be very useful. We need to be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better.
```
Comment (nylen) [inline .github/CONTRIBUTING.md:52]:
```suggestion
  - Review and test [existing PRs](https://github.com/ClassicPress/ClassicPress/pulls), especially looking at how well they fit the criteria described above. Let us know how you tested the changes and what you found. We need to be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better. Screenshots and text instructions documenting the status and testing of all PRs are very useful, as well as videos and gifs if applicable. These help us know the status and completion of each PR so that we can be thorough and careful with any changes we make, and the more eyes we can get on our PRs the better.
```

File: .github/CONTRIBUTING.md [hunk 14]
```diff
@@ -71,22 +108,75 @@ Documentation for running and updating our existing tests, as well as the code f
...
 
-Otherwise, if you're interested in backporting a change from WordPress, then please explain in a pull request or issue what the specific change is, how it works, and why it's something that makes sense for ClassicPress to adopt. Often there is a lot of good discussion about each change in the relevant WordPress Trac tickets, and it makes the maintainers' job much easier to see this discussion summarized and linked.
+### Tips for a good Backport PR.
+- Give the PRs a clear title. Explain in a pull request or issue what the specific change is, how it works and why ClassicPress in particular should adopt it. Often there is a lot of good discussion about each change in the relevant WordPress Trac tickets. It makes the maintainers' job much easier to see this discussion summarized and linked.
+- Explain what testing has been done on the PR and what may be left to do. Screenshots, gif, video and text instructions documenting the tests can be very useful. Those tell other contributors looking at the PR how to verify the changes. "This was tested in WordPress" **is not enough** to meet our review criteria, see [Review Criteria](#review-criteria) above.
```
Comment (nylen) [inline .github/CONTRIBUTING.md:123]:
```suggestion
- Explain what testing has been done on the PR and what may be left to do. Screenshots and text instructions documenting the changes and how to test them are very useful. Those tell other contributors looking at the PR how to verify the changes. "This was tested in WordPress" **is not enough** to meet our review criteria, see [Review Criteria](#review-criteria) above.
- Look through the WordPress tickets and commits associated with these changes. Make sure there are no other changes to commit, no outstanding issues with the WordPress changes mentioned in Trac tickets, and summarize your findings on your PR.
```

File: .github/CONTRIBUTING.md [hunk 15]
```diff
@@ -71,22 +108,75 @@ Documentation for running and updating our existing tests, as well as the code f
...
-Using this script for all backports saves time for you and for the maintainers, and it also uses a standardized format for commit messages which makes it possible for us to track which WordPress changes we've already included.
+3. Resolve merge conflicts (if any)
+4. Repeat steps 2 and 3
+5. Then push your branch to your fork online using:
+   
+```
+git checkout upstream/develop -B desired-branch-name
+```
+6. Use the online platform to make a PR to the `develop` branch for review.
+!["Opening a PR to the develop branch"][open_pr_br_develop]
```
Comment (nylen) [inline .github/CONTRIBUTING.md:158]: This is not the right image. Maybe let's leave that out for now?

File: .github/CONTRIBUTING.md (hunk 11 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md]: Debated it too. Committed. Thanks.

File: .github/CONTRIBUTING.md (hunk 15 above)
Comment (bahiirwa) [reply .github/CONTRIBUTING.md:157]: Will revert this. Should I squash changes?

Comment (nylen) [reply .github/CONTRIBUTING.md]: No need to squash changes, they will get rolled up into one commit when we merge this.

**Review by nylen — APPROVED** [review]
This is good to go after the tests pass. Thank you!
